File: lib/common/commandbutton.py
import logging

from .eventhandlers import CreatedEventHandler
from .uicontext import uicontext

# noinspection SpellCheckingInspection
level = 'lib.common.commandbutton'
logger = logging.getLogger('silvanus').getChild(level)


class CommandButton:

    def __init__(self, app, command):
        self.app = app
        self.command = command
        self.handlers = []
        self.created_handler = CreatedEventHandler

        self.start()

    def start(self):
        self.del_command_button()
        self.add_command_button()

    def add_command_button(self):
        self.add_control_to_panel()
        self.register_handler()

    def add_control_to_panel(self):
        if not self.command_control:
            self.scripts_panel.controls.addCommand(self.find_command_definition())

    def register_handler(self):
        handler = self.created_handler(self, self.command.PREVIEW_FLAG)
        self.handlers.append(handler)

        cmd_def = self.find_command_definition()
        if not cmd_def.commandCreated.add(handler):
            logging.error('Unable to register handler for command')
            raise RuntimeError('Unable to register handler for command')

    def stop(self):
        self.del_command_button()

    @property
    def scripts_panel(self):
        return self.app.userInterface.allToolbarPanels.itemById(self.command.PANEL_LOCATION)

    @property
    def command_definition(self):
        return self.app.userInterface.commandDefinitions.itemById(self.command.ID)

    @property
    def command_control(self):
        return self.scripts_panel.controls.itemById(self.command_definition.id) if self.command_definition else None

    def find_command_definition(self):
        return self.command_definition if self.command_definition else self.add_button_definition()

    def add_button_definition(self):
        return self.app.userInterface.commandDefinitions.addButtonDefinition(
                self.command.ID, self.command.DIALOG, self.command.DESC
        )

    def del_command_button(self):
        check_control = self.scripts_panel and self.command_definition
        control = self.command_control if check_control else None

        if self.command_definition:
            self.command_definition.deleteMe()
        if control:
            control.deleteMe()

    def execute(self, inputs):
        with uicontext(self.app, level):
            self.command.execute(inputs)

    def change(self, inputs):
        with uicontext(self.app, level):
            self.command.change(inputs)

    def validate(self, inputs):
        with uicontext(self.app, level):
            return self.command.validate(inputs)

    def create(self, inputs):
        with uicontext(self.app, level):
            self.command.create(inputs)

    def preview(self, inputs):
        with uicontext(self.app, level):
            self.command.preview(inputs)

    def destroy(self, args):
        with uicontext(self.app, level):
            self.command.destroy(args)

File: src/lib/common/commandbutton.cpp
//
//  SilvanusPro
//
//  Created by Hobbyist Maker on 7/21/20.
//  Copyright © 2020 HobbyistMaker. All rights reserved.
//

#include <Core/CoreAll.h>

#include "commandbutton.hpp"

using namespace adsk::core;
using namespace adsk::fusion;

using namespace silvanus::common;

Ptr<CommandDefinition> CommandButton::addButtonDefinition() {
    auto ui = m_app->userInterface();
    auto cmd_defs = ui->commandDefinitions();
    
    return cmd_defs->addButtonDefinition(
        m_command->getId(), m_command->getDialogName(), m_command->getDescription()
    );
}

void CommandButton::addControlToPanel() {
    Ptr<UserInterface> ui = m_app->userInterface();

    auto control = commandControl();
    if (control) {
        return;
    }

    scriptsPanel()->controls()->addCommand(findOrCreateCommandDefinition());
}

Ptr<CommandControl> CommandButton::commandControl() {
    Ptr<UserInterface> ui = m_app->userInterface();

    auto panel = scriptsPanel();
    auto cmd_def = commandDefinition();
    
    if (!cmd_def) {
        return nullptr;
    }
    
    auto controls = panel->controls();
    auto control = controls->itemById(cmd_def->id());
    
    return control;
}

Ptr<CommandDefinition> CommandButton::commandDefinition() {
    Ptr<UserInterface> ui = m_app->userInterface();
    Ptr<CommandDefinitions> cmd_defs = ui->commandDefinitions();
    Ptr<CommandDefinition> cmd_def = cmd_defs->itemById(m_command->getId());
    
    return cmd_def;
}

void CommandButton::delCommandButton() {
    auto scripts_panel = scriptsPanel();
    auto cmd_def = commandDefinition();
    auto control = commandControl();
    
    if (cmd_def) {
        cmd_def->deleteMe();
    }
    
    if (control) {
        control->deleteMe();
    }
}

Ptr<CommandDefinition> CommandButton::findOrCreateCommandDefinition() {
    auto cmd_def = commandDefinition();
    return cmd_def ? cmd_def : addButtonDefinition();
}

void CommandButton::registerHandler() {
    Ptr<UserInterface> ui = m_app->userInterface();
   
    auto cmd_def = findOrCreateCommandDefinition();
    bool is_ok = cmd_def->commandCreated()->add(m_handler.get());
    
    if (!cmd_def || !is_ok) {
        ui->messageBox("Unable to register handler for command.");
    }
}

Ptr<ToolbarPanel> CommandButton::scriptsPanel() {
    Ptr<UserInterface> ui = m_app->userInterface();
    Ptr<ToolbarPanelList> panels = ui->allToolbarPanels();
    auto item = panels->itemById(m_command->getPanelLocation());

    return panels->itemById(m_command->getPanelLocation());
}
